### ISSUE TITLE
"Failed to save link" toast but still saves

### ISSUE THREAD

**Issue by purplespider** [silverstripe/silverstripe-linkfield#349]
### Module version(s) affected

4.0.8

### Description

Upon clicking "Save" when adding any link, a toast appears stating "Failed to save link" and the modal stays on screen.

However, upon refreshing the page, the link appears in the list, so it was saved.

![Screenshot 2024-11-14 at 11 39 53](https://github.com/user-attachments/assets/ef7aea62-be26-4b2f-9bc3-c2ac11638dcf)

Works fine locally, issue only occurs once deployed. 

No errors in console and no fetch errors.

The only thing different I can see, between when it works locally, and doesn't work on production is locally, after the fetch request to `admin/linkfield/linkForm/0` there are two more fetch requests (to `admin/linkfield/linkForm/13/field/PageID/tree` and `admin/linkfield/data`) netiher of which appear when doing a save on production.

Things tried:
* Different PHP versions
* Flushing & dev/build

### How to reproduce

```php
private static $has_many = [
     'HeaderButtons' => Link::class . '.Owner',
];

//...
    $fields->addFieldToTab('Root.Main', MultiLinkField::create('HeaderButtons'));
//...
````

### Validations

- [X] Check that there isn't already an issue that reports the same bug
- [X] Double check that your reproduction steps work in a fresh installation of [`silverstripe/installer`](https://github.com/silverstripe/silverstripe-installer) (with any code examples you've provided)

**Comment by GuySartorelli** [comment]
> Works fine locally, issue only occurs once deployed.

Can you please try to find a way to reproduce this locally? If we can't reproduce it, we can't fix it. Clearly the information you provided in "How to reproduce" won't be sufficient on its own.

**Comment by purplespider** [comment]
Yeah, I appreciate that, Guy. With no errors in the console, I'm a bit stumped, though. I was hoping someone familiar with the module may have some ideas for areas to investigate that could be causing the "Failed" message to trigger. Thanks.

**Comment by lerni** [comment]
If a Webserver appends `; charset=utf-8` to JSON response headers, it exposes this bug in JS code that does exact Content-Type matching. To reproduce locally with DDEV add `.ddev/apache/charset-json.conf`
```
<IfModule mod_headers.c>
    # When response Content-Type is exactly "application/json",
    # rewrite it to include charset — matching common production server behaviour
    Header edit "Content-Type" "^application/json$" "application/json; charset=utf-8"
    Header onsuccess edit "Content-Type" "^application/json$" "application/json; charset=utf-8"
</IfModule>
```
The root cause is in `silverstripe/admin`, specifically in `client/src/lib/Backend.js`, in the `decode()` function:
```js
function decode(contentType, text) {
  switch (contentType) {
    case 'application/json':   // ← exact match only!
      return JSON.parse(text);
    default:
      throw new Error(`Can't decode format: ${contentType}`);
  }
}
```
The `decode()` switch uses **exact string matching**, so `"application/json; charset=utf-8" !== "application/json"` → falls through to `default` → throws.

### Chain of events:
1. POST to save link → succeeds server-side (link IS saved)
2. Response comes back with `Content-Type: application/json; charset=utf-8`
3. `Backend.js decode()` throws `Can't decode format: application/json; charset=utf-8`
4. Error propagates up to `LinkModal.onSubmit` catch block
5. Catch block shows "Failed to save link" toast, swallows the error silently
6. `onSuccess()` never fires → no follow-up GET requests → modal stays open
7. DDEV's Apache returns bare `application/json` → works locally

**Comment by purplespider** [comment]
@lerni Great work narrowing that down! I can confirm your fix works me. Thank you!

**Comment by lerni** [comment]
With SS 6.2.x this is fixed with https://github.com/silverstripe/silverstripe-admin/pull/2140 and released with [silverstripe-admin 3.2.1](https://github.com/silverstripe/silverstripe-admin/releases/tag/3.2.1)
